File: packages/sdk/python/examples/simple_agent.py
import asyncio
import os
from solders.keypair import Keypair
from x402_upl import SolanaX402Client, X402Config, ServiceDiscovery


async def main():
    private_key_path = os.getenv("X402_WALLET_PATH", "wallet.json")

    try:
        import json
        with open(private_key_path, "r") as f:
            secret_key = json.load(f)
            wallet = Keypair.from_bytes(bytes(secret_key))
    except FileNotFoundError:
        print("Wallet not found, generating new wallet...")
        wallet = Keypair()
        print(f"New wallet address: {wallet.pubkey()}")
        print("Fund this wallet with SOL to use the SDK")

    config = X402Config(
        network="devnet",
        rpc_url="https://api.devnet.solana.com",
    )

    client = SolanaX402Client(wallet, config)
    discovery = ServiceDiscovery()

    try:
        balance = await client.get_balance()
        print(f"Wallet: {client.get_wallet_address()}")
        print(f"Balance: {balance} SOL")

        print("\nDiscovering x402 services...")
        services = await discovery.discover(
            category="AI & ML",
            max_price=0.10,
            limit=5
        )

        print(f"Found {len(services)} services:")
        for i, service in enumerate(services, 1):
            print(f"\n{i}. {service.name}")
            print(f"   Description: {service.description}")
            print(f"   Endpoint: {service.resource}")
            if service.pricing:
                print(f"   Price: {service.pricing.get('amount')} {service.pricing.get('asset')}")

        if services:
            print("\n" + "="*50)
            print("Testing first service...")
            service = services[0]

            try:
                result = await client.get(service.resource)
                print("Service response:")
                print(result)

            except Exception as e:
                print(f"Service call failed: {e}")

    finally:
        await client.close()
        await discovery.close()


if __name__ == "__main__":
    asyncio.run(main())

File: packages/sdk/rust/examples/simple_agent.rs
use solana_sdk::signature::Keypair;
use x402_upl::{SolanaX402Client, X402Config, ServiceDiscovery, DiscoveryConfig};

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    println!("x402-upl Rust SDK - Simple Agent Example\n");

    let wallet = Keypair::new();
    println!("Generated wallet: {}", wallet.pubkey());

    let config = X402Config {
        network: "devnet".to_string(),
        rpc_url: Some("https://api.devnet.solana.com".to_string()),
        ..Default::default()
    };

    let client = SolanaX402Client::new(wallet, config)?;

    let balance = client.get_balance("SOL")?;
    println!("Wallet balance: {} SOL\n", balance);

    if balance == 0.0 {
        println!("⚠️  Wallet has no balance. Fund it to test payments:");
        println!("   solana airdrop 1 {} --url devnet\n", client.get_wallet_address());
    }

    println!("Discovering x402 services...");
    let discovery_config = DiscoveryConfig::default();
    let discovery = ServiceDiscovery::new(discovery_config)?;

    let services = discovery
        .discover(
            Some("AI"),
            None,
            Some(0.10),
            None,
            None,
            5
        )
        .await?;

    println!("Found {} services:\n", services.len());

    for (i, service) in services.iter().enumerate() {
        println!("{}. {}", i + 1, service.name);
        println!("   Description: {}", service.description);
        println!("   Endpoint: {}", service.resource);
        if let Some(pricing) = &service.pricing {
            println!("   Price: {} {}", pricing.amount, pricing.asset);
        }
        println!();
    }

    if let Some(service) = services.first() {
        println!("Testing service: {}", service.name);

        match client.get::<serde_json::Value>(&service.resource, None).await {
            Ok(response) => {
                println!("✓ Service response:");
                println!("{}", serde_json::to_string_pretty(&response)?);
            }
            Err(e) => {
                println!("✗ Service call failed: {}", e);
            }
        }
    }

    Ok(())
}
